vllm/entrypoints/openai/serving_chat.py

```diff
@@ -356,7 +359,10 @@ async def chat_completion_full_generator(
 
         return response
 
-    def _load_chat_template(self, chat_template: Optional[str]):
+    async def _load_chat_template(self, chat_template: Optional[str]):
+        while self.tokenizer is None:
```

Comment:
I think this line is faulty. There is no code that actually initializes `self.tokenizer = None` before `self._load_chat_template` is called. You might want to instead use `hasattr` to check the existence of the attribute.

Reply:
I could open another PR simply replacing `while self.tokenizer is None` by `while getattr(self, "tokenizer", None) is None`.

Reply:
I guess inside `tests/async_engine/test_chat_template.py`, you can use another `MockServingChat` that doesn't have the `chat_template` attribute. However, I am doubtful about the value of such a test since it does not depend on `OpenAIServing.__init__`, making it useless if another commit changes the logic.

IMO It would be more useful to test a situation where the tokenizer takes considerably longer to load, making it more likely that the chat template will be accessed before the engine has fully loaded.

Reply:
Honestly, I think the current design is not that great as it puts `async` logic into `__init__`. It would be better if `__init__` requires `tokenizer` and `chat_template` upfront which would force users to place the `async` logic outside of the constructor.

Reply:
I agree with that sentiment. The only reason `_post_init` is async in the first place is that `engine.get_model_config` is async and in turn this is only async in order to enable the engine-workers to use ray. So 95% of the code is already synchronous and the remaining 5% are only artificially asynchronous to enable ray workers.

Reply:
Loading the chat_template used to be synchronous (before this PR) but this didn't mesh with the async code in the `ServingEngine`'s `__init__`

Reply:
> Honestly, I think the current design is not that great as it puts async logic into __init__. It would be better if __init__ requires tokenizer and chat_template upfront so that developers are encouraged to place the async logic outside of the constructor.
> 
> To maintain the same functionality as the current __init__, we can have a separate async staticmethod factory that does both async and initialization.

If you don't mind, I'll work on a PR regarding this issue. This should make it no longer necessary to test the async behaviour of `_load_chat_template`, so the relevant tests will be removed.

Reply:
Feel free to open a PR, but I'm not sure what you mean regarding the tests. As long as `OpenAIServingChat` can still be initiated in an async function I'm fine with everything ;)

Reply:
I mean that the chat template tests will be reverted to sync functions as opposed to async.